### ISSUE TITLE
[Priority Expander] CA Priority Expander scales up both high & low value ASGs triggered by ONE pod. 

### ISSUE THREAD

**Issue by xinxingliu90** [kubernetes/autoscaler#5655]
Hi, my team created `cluster-autoscaler-priority-expander` config map with two ASGs: 
- ASG-1   (high priority)  
- ASG-2   (low priority)

My config map looks like this:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-autoscaler-priority-expander
  namespace: autoscaler-cloud-resource-management
data:
  priorities: |-
    10: 
      - <ASG-2>
    50: 
      - <ASG-1>
```
and I have put `- --expander=priority` in the CA. 

**Expect behavior:** 
I expect `ASG-1` to scale up first and `ASG-2` scale up when `ASG-1` is 100% scaled up. 

**Problematic behavior**
But when I submit ONE pod, both ASGs scale up! 

_( Note: I am using CA image: k8s.gcr.io/autoscaling/cluster-autoscaler:v1.23.1 )_

Could anyone help us find the cause why Priority Expander scales up All ASGs instead scale up the high-priority ASG?

**Comment by xinxingliu90** [comment]
More details to help you triage - here is the CA log:
[CA_log.txt](https://github.com/kubernetes/autoscaler/files/11154039/CA_log.txt)

To summarize the above logs:

1. Priority Expander chose the high-priority `ASG-1` ( we name it as`bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y`) "the highest available" and scale from 0 --> 1.
```
I0404 22:41:52.271370       1 priority.go:114] Successfully loaded priority configuration from configmap.
I0404 22:41:52.271383       1 priority.go:163] priority expander: bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y chosen as the highest available
```

2. Then Pod  `bolt-qnsp5ye67g-djphkqw9ri` "marked as unschedulable can be scheduled."
```
I0404 22:42:52.863474       1 filter_out_schedulable.go:157] Pod turi-bolt.bolt-qnsp5ye67g-djphkqw9ri marked as unschedulable can be scheduled on node template-node-for-bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y-6719190960263234142-upcoming-0. Ignoring in scale up.
I0404 22:42:52.863493       1 filter_out_schedulable.go:170] 0 pods were kept as unschedulable based on caching
I0404 22:42:52.863502       1 filter_out_schedulable.go:171] 1 pods marked as unschedulable can be scheduled.
```

3. At last, low-priority `ASG-2` (we name it `bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-zcz1sdsugb` )is "chosen as the highest available" and scale from 0 --> 1.  (WHY ??)
```
I0404 22:43:53.295326       1 priority.go:163] priority expander: bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-zcz1sdsugb chosen as the highest available
I0404 22:43:53.295332       1 scale_up.go:468] Best option to resize: bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-zcz1sdsugb
I0404 22:43:53.295340       1 scale_up.go:472] Estimated 1 nodes needed in bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-zcz1sdsugb
I0404 22:43:53.295361       1 scale_up.go:595] Final scale-up plan: [{bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-zcz1sdsugb 0->1 (max: 5)}]
```

**Comment by MaciekPytel** [comment]
CA scaled-up the second ASG after the scale-up in first one finished and the pod still couldn't schedule. In logs immediately before the second scale-up:

```I0404 22:43:53.294457       1 clusterstate.go:248] Scale up in group bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y finished successfully in 2m0.444665932s
...
I0404 22:43:53.294683       1 klogx.go:86] Pod turi-bolt/bolt-qnsp5ye67g-djphkqw9ri is unschedulable
I0404 22:43:53.294814       1 scale_up.go:376] Upcoming 0 nodes
```

Based on this the sequence of events was:
1. CA noticed a pending pod
2. It triggered scale-up in first ASG as per priority expander configuration
3. The scale-up finished (meaning the new node became ready), but the pod couldn't schedule on new node because fuse isn't ready: 
```
I0404 22:43:53.294902       1 scale_up.go:300] Pod bolt-qnsp5ye67g-djphkqw9ri can't be scheduled on bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y, predicate checking error: Insufficient devices.applecloud.io/fuse_v2; predicateName=NodeResourcesFit; reasons: Insufficient devices.applecloud.io/fuse_v2; debugInfo=
```
4. As far as CA knows the node has already finished booting (kubernetes status conditions show it is ready) and it apparently can't run the pending pod, so CA decided to try the other ASG.

This isn't really related to priority expander. It's a known problem between CA and device plugins. There are currently two ways of dealing with this:
1. Implement your own [CustomResourceProcessor](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/processors/customresources/custom_resources_processor.go) or just extend the GPU one by adding a check for your own label denoting nodes with fuse.
2. Use ignore-taint (either via flag or taint prefix). The name of ignore-taint is misleading, it's really only meant as a startup taint intended specifically for solving the exact issue you're facing. Any node that has ignore-taint will be considered still booting up, so if you create your node with ignore-taint and remove it once your fuse is ready it should solve the issue above. The caveats is that you _must_ remove the taint within 15 minute of scale-up. Too many long-running nodes with ignore-taints will cause CA to completely stop working and it won't be easy to debug (no obvious error messages).

I would personally recommend option 1. Option 2 may be easier to deploy, but I personally think it's too much of a footgun. I've seen so many CA outages caused by misusing ignore-taint (including by very large and advanced users) that I hesitate to even mention it as an option anymore.

**Comment by xinxingliu90** [comment]
Thanks @MaciekPytel ! I forgot to mention that the pod scheduled successfully onto `ASG-1` as soon as it scaled up. You pointed out the following log line:
```
I0404 22:43:53.294902       1 scale_up.go:300] Pod bolt-qnsp5ye67g-djphkqw9ri can't be scheduled on bolt-aws1-dev-cloud_resource_management-r6i.8xlarge-integration-az1_2_3-kl1vjsfh3y, predicate checking error: Insufficient devices.applecloud.io/fuse_v2; predicateName=NodeResourcesFit; reasons: Insufficient devices.applecloud.io/fuse_v2; debugInfo=
```
you remind of that because my pod request more resources other than `cpu` and `gpu`, therefore in pod spec, I follow CA document: 

```
If your pods request resources other than `cpu` and `memory`, you need to tag ASG with key `k8s.io/cluster-autoscaler/node-template/resources/`.
```
and I added ASG tag `k8s.io/cluster-autoscaler/node-template/resources/devices.applecloud.io/fuse_v2`.

**My suspicion is:**  Priority Expander does not take the tag  `k8s.io/cluster-autoscaler/node-template/resources/devices.applecloud.io/fuse_v2` into consideration when “filter_out_schedulable” , so it thinks ZERO pod is schedulable therefore triggers scale up another “low-priority” ASG. ( Even though actually, one pod is schedulable and is scheduled to the “high-priority” ASG and start running later) 

Do you think my guess above ⬆️ could be the reason?

**Comment by xinxingliu90** [comment]
@MaciekPytel I think now I understand your solution and I will try both to see if it can solve the race condition.
